src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java

```diff
@@ -84,21 +93,31 @@ public Collection<S3Key> getKeysForSite(Integer siteId) {
                 .collect(Collectors.toList());
     }
 
-    public S3Key getEncryptionKeyForSite(Integer siteId) {
-        Collection<S3Key> keys = getKeysForSite(siteId);
+    public List<S3Key> getKeys(int siteId) {
+        return siteToKeysMap.getOrDefault(siteId, new ArrayList<>());
+    }
+
+    public List<S3Key> getMostRecentKeysForSite(int siteId) {
+        List<S3Key> keys = getKeys(siteId);
         if (keys.isEmpty()) {
+            throw new IllegalStateException("No keys found for site ID: " + siteId);
+        }
+        return keys.stream()
+                .sorted(Comparator.comparingLong(S3Key::getCreated).reversed())
+                // Get most recently created 3 keys
+                .limit(3)
+                .collect(Collectors.toList());
+    }
+
+     public S3Key getEncryptionKeyForSite(Integer siteId) {
```

Comment:
I think we can combine this with the function above, and always send all active keys.